### PR TITLE
Add request deadline

### PR DESCRIPTION
**Pull request by pepone**
This PR updates Ice2 request header to include the request deadline, the deadline is encoded as the number of milliseconds since epoch, we use 0 to represent an infinite deadline and it gets translated to DateTime.MaxValue on C#.  The deadline is encoded using the varlong type.  Using 0 as the infinite deadline is just a small optimization, the more natural would be -1 (infinite timespan) but using 0 save a few bytes with the varlong encoding.

### CODE REVIEW

File: csharp/src/Ice/Current.cs [hunk 1]
```diff
@@ -18,6 +19,10 @@ public class Current
         /// <summary>The <see cref="Connection"/> over which the request was dispatched.</summary>
         public Connection Connection { get; }
 
+        /// <summary>The request deadline. The peer sets the deadline to the absolute time at which its timeout will be
+        /// triggered. With Ice1 the peer doesn't send a deadline and it always has the default value.</summary>
```
Comment (bernardnormier) [inline csharp/src/Ice/Current.cs:23]: For ice1, I think it should be MaxValue, not default.

Comment (bernardnormier) [reply csharp/src/Ice/Current.cs:23]: Also, it should be sorted correctly - after Current.

File: slice/Ice/Ice2Definitions.ice [hunk 2]
```diff
@@ -40,6 +40,7 @@ module Ice
         string operation;
         bool? \idempotent;       // null equivalent to false
         Priority? priority;      // null equivalent to 0
+        varlong deadline;
```
Comment (bernardnormier) [inline slice/Ice/Ice2Definitions.ice:43]: varlong suggests we use negative values - typically to use "-1" to mean infinite. If we don't, it should be varulong.

File: csharp/src/Ice/OutgoingRequestFrame.cs [hunk 3]
```diff
@@ -355,7 +362,14 @@ internal override void Finish()
             else
             {
                 Debug.Assert(Protocol == Protocol.Ice2);
-                ostr.WriteIce2RequestHeaderBody(Identity, Facet, Location, Operation, IsIdempotent);
+                ostr.WriteIce2RequestHeaderBody(Identity,
+                                                Facet,
+                                                Location,
+                                                Operation,
+                                                IsIdempotent,
+                                                // An infinite invocation timeout is encoded as 0
```
Comment (bernardnormier) [inline csharp/src/Ice/OutgoingRequestFrame.cs:370]: I would use `-1`, which is the value for InfiniteTimeSpan.

**Review by bernardnormier — COMMENTED** [review]
I would use `-1`. Not sure why you think it uses more bytes.

**Comment by pepone** [comment]
> I would use `-1`. Not sure why you think it uses more bytes.

I confused our var encoding already optimizes the length

File: csharp/src/Ice/Current.cs (hunk 1 above)
Comment (pepone) [reply csharp/src/Ice/Current.cs]: Fixed

File: slice/Ice/Ice2Definitions.ice (hunk 2 above)
Comment (pepone) [reply slice/Ice/Ice2Definitions.ice:43]: keep it as varlong, as we will use -1 for infinite deadline

File: csharp/src/Ice/OutgoingRequestFrame.cs (hunk 3 above)
Comment (pepone) [reply csharp/src/Ice/OutgoingRequestFrame.cs]: Fixed

File: csharp/src/Ice/Current.cs [hunk 4]
```diff
@@ -21,6 +22,11 @@ public class Current
         /// <summary>The request context, as received from the client.</summary>
         public Dictionary<string, string> Context => IncomingRequestFrame.Context;
 
+        /// <summary>The request deadline. The peer sets the deadline to the absolute time at which its invocation
```
Comment (bernardnormier) [inline csharp/src/Ice/Current.cs:25]: What about:

The deadline corresponds to the request's expiration time. Once the deadline is reached, the caller is no longer interested in the response and discards the request. The server-side runtime does not enforce this deadline - it's provided "for information" to the application. The Ice client runtime sets this deadline automatically using the proxy's invocation timeout and sends it with ice2 requests but not with ice1 requests. As a result, the deadline for an ice1 request is always DateTime.MaxValue on the server-side even though the invocation timeout is usually not infinite.

[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/Current.cs:25]: mm maybe remove "absolute" here, as this is a DateTime object?

File: csharp/src/Ice/Ice2Definitions.cs [hunk 5]
```diff
@@ -29,7 +30,8 @@ internal enum FrameType : byte
             string facet,
             IReadOnlyList<string> location,
             string operation,
-            bool idempotent)
+            bool idempotent,
+            TimeSpan deadline)
```
Comment (bernardnormier) [inline csharp/src/Ice/Ice2Definitions.cs:34]: I would pass directly the Deadline date time here and not introduce a confusing intermediary TimeSpan deadline parameter.

File: csharp/src/Ice/Ice2Definitions.cs [hunk 6]
```diff
@@ -65,6 +67,8 @@ internal enum FrameType : byte
             }
 
             bitSequence[3] = false; // TODO: source for priority.
+
+            ostr.WriteVarLong((long)deadline.TotalMilliseconds);
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/Ice2Definitions.cs:71]:
```
// An infinite deadline is encoded as -1
            ostr.WriteVarLong(
                deadline == DateTime.MaxValue ? -1 : (long)(deadline - DateTime.UnixEpoch).TotalMilliseconds);
```
something like that? as per your other comment, we should pass the DateTime deadline

File: csharp/src/Ice/IncomingRequestFrame.cs [hunk 7]
```diff
@@ -62,6 +69,10 @@ public IncomingRequestFrame(Protocol protocol, ArraySegment<byte> data, int size
                 Operation = requestHeaderBody.Operation;
                 IsIdempotent = requestHeaderBody.Idempotent ?? false;
                 Priority = requestHeaderBody.Priority ?? default;
+                var deadline = TimeSpan.FromMilliseconds(requestHeaderBody.Deadline);
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/IncomingRequestFrame.cs:72]: Yes we should validate it

File: csharp/src/Ice/Current.cs (hunk 4 above)
Comment (pepone) [reply csharp/src/Ice/Current.cs]: Fixed as suggested

File: csharp/src/Ice/Ice2Definitions.cs (hunk 5 above)
Comment (pepone) [reply csharp/src/Ice/Ice2Definitions.cs]: Fixed

File: csharp/test/Ice/timeout/AllTests.cs [hunk 8]
```diff
@@ -186,6 +186,36 @@ public static void RunWithController(TestHelper helper, IControllerPrx controlle
...
+            if (!ice1)
+            {
+                output.Write("testing deadlines... ");
+                output.Flush();
+                {
+                    var comm1 = new Communicator(
+                        communicator.GetProperties(),
+                        invocationInterceptors: new InvocationInterceptor[]
+                        {
+                        (target, request, next, cancel) =>
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/test/Ice/timeout/AllTests.cs]: Fixed

File: csharp/src/Ice/IncomingRequestFrame.cs [hunk 9]
```diff
@@ -62,6 +69,10 @@ public IncomingRequestFrame(Protocol protocol, ArraySegment<byte> data, int size
...
+                // The infinite invocation timeout is encoded as 0 and converted to DateTime.MaxValue
```
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/IncomingRequestFrame.cs]: Fixed

File: csharp/src/Ice/Current.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pepone) [reply csharp/src/Ice/Current.cs]: Fixed with Bernard's suggestion

**Review by bernardnormier — APPROVED** [review]
Looks good.

**Pull request merged**